File: simpleast-core/src/main/java/com/agarron/simpleast_core/simple/SimpleRenderer.java
package com.agarron.simpleast_core.simple;

import android.content.Context;
import android.support.annotation.StringRes;
import android.text.SpannableStringBuilder;
import android.widget.TextView;

import com.agarron.simpleast_core.builder.Parser;
import com.agarron.simpleast_core.node.Node;
import com.agarron.simpleast_core.renderer.SpannableRenderer;

import java.util.Collection;

public class SimpleRenderer {

    public static void renderBasicMarkdown(final TextView textView, @StringRes final int sourceResId) {
        final CharSequence source = textView.getContext().getString(sourceResId);
        renderBasicMarkdown(textView, source);
    }

    public static void renderBasicMarkdown(final TextView textView, final CharSequence source) {
        textView.setText(renderBasicMarkdown(source, textView.getContext()));
    }

    public static SpannableStringBuilder renderBasicMarkdown(final CharSequence source, final Context context) {
        return render(source, SimpleMarkdownRules.createSimpleMarkdownRules(), context);
    }

    public static SpannableStringBuilder render(final CharSequence source, final Collection<Parser.Rule<Node>> rules, final Context context) {
        final Parser<Node> parser = new Parser<>();
        for (final Parser.Rule<Node> rule : rules) {
            parser.addRule(rule);
        }

        return SpannableRenderer.render(new SpannableStringBuilder(), parser.parse(source, false), context);
    }
}

File: simpleast-core/src/main/java/com/agarron/simpleast_core/SimpleRenderer.kt
package com.agarron.simpleastcore

import android.text.Spannable
import android.text.SpannableStringBuilder

fun render(ast: Collection<Node>): CharSequence {
    val builder = SpannableStringBuilder()
    ast.forEach { renderNode(it, builder) }
    return builder
}

fun renderNode(node: Node, builder: SpannableStringBuilder) {
    if (node is StyleNode){
        val startIndex = builder.length
        node.getChildren().forEach { renderNode(it, builder) }
        node.styles.forEach { builder.setSpan(it, startIndex, builder.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE) }
    } else if (node is TextNode) {
        builder.append(node.content)
    } else {
        throw IllegalArgumentException("invalid node of type: " + node.type)
    }
}
